### ISSUE TITLE
Wrong result after decompression.

### ISSUE THREAD

**Issue by fengdh** [abraidwood/minilzo-js#1]
In order to decompress LZO compressed file inside browser, I have forked your minilzo-js to accept and output data in Uint8Array for decompression. It works well except fail on a small part of my test data (among 73 data, only 2 failed).

You can view my work on https://github.com/fengdh/minilzo-decompress.js, including a sample LZO compressed binary data file which will produce a wrong result (expecting size of 32754, not 32737). Here is an online demo on jsbin (http://jsbin.com/xuqona/edit?html,js,console,output). You have to download lzo_compressed_sample.data file first and choose it from local, then view result as console output .

Would you help me to figure out the problem? Thank you.

**Comment by abraidwood** [comment]
It's been quite a while since I looked at any of this code, but I will try to remember and to help.

A couple of questions:
1. Does the file decompress as expected with the minilzo decompress functions in the version I wrote?
2. How did you compress the file?  If it's with my code, then there's a chance that the compression code is broken.

**Comment by fengdh** [comment]
Thanks for your reply.

Response to your question:
1. Most of files (71 out 73) are decompressed as expected with minilzo.js, but only 2 files failed. Your version of decompress function takes string as input/output, so I have to modified it to work on Uint8Array. My port should have no significant changes inside body of decompress() function.
2. I did not compress those files, which are LZO compressed data blocks extracted from a dictionary file (*.mdx of MDict). It must be compressed correctly with C library, since the MDict Viewer can handle it well. I guess maybe there is a minor bug inside decompress() function.

**Comment by fengdh** [comment]
Look inside failed files, I found that if first 2-byte of compressed data equals to [0x12, 0x00]  or [18, 0] in decimal. decompress() will execute like below:
- start with ip = 0

```
      if (pInBuf[ip] > 17) {
        t = pInBuf[ip++] - 17;   // <-- t = 1 at this point
        if (t < 4) {
          // goto match_next
          outerLoopPos = 2;
          innerLoopPos = 3;   // <-- and initialized as left
```
- then

``` javascript
        if(outerLoopPos === 2) {
          outerLoopPos = 0;
          innerLoopPos = 0; // this is a bug   // <-- comment by abraidwood in original minilzo-js
```
- goto 

```
            if(innerLoopPos === 0) {
              if (t >= 64) { .. // skip
              } else if (t>=32) { ... // skip
              } else if (t>=16) { ... // skip
              } else {
                m_pos = op - 1;
                m_pos -= t >> 2;
                m_pos -= pInBuf[ip++] << 2;    // <-- got m_pos = -1, IS IT A BUG?

                out = buf.require(2);
                out[op++] = out[m_pos++];      // <-- copy out of beginning of data
                out[op++] = out[m_pos];
```

I wonder if it is a bug?

Another failed file turn to be corrupted partly in the middle of decompression. Maybe somewhere shift operation caused overflow when calculating m_pos?

**Comment by abraidwood** [comment]
Yes, that looks like a bug.  I'll see what I can do about it.

**Comment by abraidwood** [comment]
I'm trying to decode the lzo_compressed_sample.data file that's in your repo, but don't have the uncompressed sample to compare my result against.

Are you able to add the correctly uncompressed data to your repo?

**Comment by fengdh** [comment]
I have upload the uncompressed sample data to my repo (https://github.com/fengdh/minilzo-decompress.js/blob/master/lzo_decompressed_sample.data).

It is decompressed with java-compress (https://code.google.com/p/java-compress/).

Result data should consist entries of {offset, keyword}, which is part of dictionary keyword index table:

offset[0]    4-byte, big-endian int
keyword[0]   UTF-16LE encoding string, null-terminated
...
offset[n]
keyword[n]

In the decompressed data, the first entry is {offset: 0, keyword: "24-7"}, and the last one is {offset:  0x0083A352, keyword: "aftermath"}.

Sample compressed data #2 starts with [0x18, 0x00] or [24, 0] in decimal, which minilzo-js decompresses to a half corrupted result.

I also ported  MiniLZO#lzo1x_decompress(..) in java-compress to JavaScript, which can decompress both of my sample data correctly (now updated to my repo as minilzo-decompress2.js).

**Comment by abraidwood** [comment]
Thanks for the decompressed version and the additional detail.

I've added a new https://github.com/abraidwood/minilzo-js/blob/master/lzo1x_decompress_safe.js which decompresses that file successfully and appears faster than the one based on the java-compress implementation.  It's a new reimplementation of lzo_decompress_safe from minilzo.c version 2.09.

I'm keen for you to try it with your other test files.

Note that the code is very messy because I've not done any cleanup or implemented the input/output overrun checks that exist in lzo_decompress_safe in minilzo.c.

**Comment by fengdh** [comment]
I have tried with lzo1x_decompress_safe.js, but it failed every other data block except the first one (same as lzo_decompressed_sample.data). It breaks the for(;;) loop too soon at line #281 where (match_done() === 0) is satisfied.

My port of java-compress did decompress all the data block as expected. Maybe I have to rely on it and make it a bitter fast.

**Comment by abraidwood** [comment]
I have found and corrected another error around memory allocation.

I now have a large set of sample files (23000 files from my local source folders) which I am testing against.  I am testing both my version and your java-compress port decompressing files created with an executable based on the compress side of the testmini.c code in the minilzo distribution.

Currently, 50 files are failing with both decompression implementations, 20 files are failing only in mine, and 1 only in yours.  I am investigating why as it's possibly a compression bug.

**Comment by abraidwood** [comment]
I've corrected an additional memory allocation error and now lzo1x_decompress_safe.js passes all my tests.

I think there's an allocation error in your version too, but have not yet tracked it down.

Edit: pull request submitted to your version

**Comment by abraidwood** [comment]
I've updated the base lzo1x port of the minilzo1x functions.  Both compress and decompress operate on UInt8Arrays.  The syntax has changed & is now a little more similar to those of the original.

I've tested against ~25k files of various types (mainly text files with some images, sound files, jars, object files, etc) from my development directories checking the js compression against the minilzo.c compressed version o the same file, and the js decompression of the minilzo.c compressed version against the original.

Performance seems quite good - 30-40MByte/s for both compress and decompress on my 2.5GHz i5 (IvyBridge) macbook pro - that's about twice as good as the old version.

Further optimisation is definitely possible on the compression side as I've not implemented the 4-byte copies present in the decompressor.
